### DISCUSSION TITLE
Getting page info before client side navigation is complete

### DISCUSSION THREAD

**Discussion by lmatteis** [vercel/next.js#41255] [Help]
I have a simple navigation bar with `<Link />`s to other pages. All pages are `getServerSideProps`.

Each time a page is clicked, the user waits for the .json to download and then renders the new page.

We would like to instead show some feedback immediately as the .json is being requested. For instance we'd like to highlight the currently selected page in the navigation to give feedback to the user that they selected that page.

Problem is that `useRouter()` doesn't return the "currently fetching page" but only returns the page once fully available. So there's a bit of lag on the user-interaction part of it.

We can't use shallow routing since these are different pages (different urls). Is there a way to do this using just native page navigation?

**Comment by gaston-flores** [comment]
You can use [router events](https://nextjs.org/docs/api-reference/next/router#routerevents) to be notified when a new route is being loaded. Something like:
```ts
import { useRouter } from "next/router";
import { useEffect, useState } from "react";

function useRouteLoading() {
  const router = useRouter();
  const [routeLoading, setRouteLoading] = useState(null);
  useEffect(() => {
    const handleStart = (url) => {
      setRouteLoading(url);
    };

    const handleStop = () => {
      setRouteLoading(null);
    };

    router.events.on("routeChangeStart", handleStart);
    router.events.on("routeChangeComplete", handleStop);
    router.events.on("routeChangeError", handleStop);

    return () => {
      router.events.off("routeChangeStart", handleStart);
      router.events.off("routeChangeComplete", handleStop);
      router.events.off("routeChangeError", handleStop);
    };
  }, [router]);
  return routeLoading;
}

export default useRouteLoading;
```
[Example](https://codesandbox.io/s/next-use-route-loading-46zgtw?file=/useRouteLoading.js).

**Comment by lmatteis** [comment]
@gaston-flores thanks that actually works. But was wondering if there's a more native way to make it work. Because now I have a bunch of components that call `useRouter()` to understand what the current page is. With this ad-hoc useRouteLoading I'd have to have my navigation check both useRouter() and useRouteLoading() to get the "currently selected page" information.

I guess that would work but feels a bit hacky. 

I thought there would be a way in Next to simply specify whether the page is "blocking" or not - where you either want your components to wait until all the page data is there, or whether you'd want to to immediately transition and show pending states until the data is there. Seems like specifiying either of these should be an option in NextJS instead of having to resort to a custom hook.

EDIT: this might be a similar issue: https://github.com/vercel/next.js/discussions/32243
